### PR TITLE
Pull request for the Auton testing branch

### PR DESCRIPTION
**Pull request by SJJCoding**
This pull request contains working code for centering the robot on the cone stack as well as some constant changes. 

This code adds a method that drives the robot towards the cone stack (provided that it sees it), and turns faster while the bot is farther away from the stack, and strafes more when it is closer. 

This method has been tested at the shop and functions.

### CODE REVIEW

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java [hunk 1]
```diff
@@ -286,12 +301,23 @@ public double coneStackWidthMotorPower(double coneStackWidth) {
...
+    /**
+     * calculates the strafing power for the robot drivetrain based on the pixel offset from the cone stack and the width of the cone stack
+     * @param pixelOffset how far cone stack is horizontally from center of camera field of view in pixels
+     * @param width how wide the cone stack is in the camera frame in pixels
+     * @return strafing power for robot drivetrain
      */
-    public double coneStackPixelsMotorPower(double pixelOffset) {
-        return Constants.CONE_STACK_CENTERING_KP * pixelOffset;
+    public double coneStackPixelsAndWidthToStrafingPower(double pixelOffset, double width) {
+        return (Constants.CONE_STACK_CENTERING_PROPORTIONAL_KP * pixelOffset)*((1.0/200.0)*width);
```
Comment (SJJCoding) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java:320]: Are these magic numbers? They are part of an equation and are only used once

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java [hunk 2]
```diff
@@ -240,6 +248,11 @@ public void centerJunctionTop(GrabberCameraPipeline pipeline) {
...
     }
 
+    /**
+     * uses the robot camera to guide the robot so it can drive forward while centering on the stack
+     * @param pipeline the RobotCameraPipeline being used by the robot camera
+     */
+
     /**
      * uses the robot camera to guide the robot so it can drive forward while centering on the stack
      * @param pipeline the RobotCameraPipeline being used by the robot camera
```
Comment (malavika1) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java:258]: You don't need the same comment twice

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java [hunk 3]
```diff
@@ -248,20 +261,22 @@ public void centerConeStack(RobotCameraPipeline pipeline) {
         double xOffset, width;
 
         do {
-            xOffset = pipeline.xPosition - Constants.CAMERA_CENTER_X;
+            xOffset = 400-pipeline.xPosition;
```
Comment (malavika1) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java:264]: Where is this 400 coming from? This should also be a variable with a useful name

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java [hunk 4]
```diff
@@ -248,20 +261,22 @@ public void centerConeStack(RobotCameraPipeline pipeline) {
...
             width = pipeline.width;
 
             // drive forward while centering on the cone stack if contour exists
             if (width == 0) {
                 break;
             } else {
-                driveWithIMU(coneStackPixelsMotorPower(xOffset), coneStackWidthMotorPower(width), 0.0);
-                telemetry.addData("xMotorPower", coneStackPixelsMotorPower(xOffset));
+                //driveWithIMU(coneStackPixelsAndWidthToStrafingPower(xOffset, width), coneStackWidthMotorPower(width), 1.09*coneStackPixelsAndWidthToTurningPower(xOffset, width));
```
Comment (malavika1) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java:271]: No commented out code

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java [hunk 5]
```diff
@@ -248,20 +261,22 @@ public void centerConeStack(RobotCameraPipeline pipeline) {
...
+                telemetry.addData("xStrafingPower", coneStackPixelsAndWidthToStrafingPower(xOffset, width));
+                telemetry.addData("xTurningPower", coneStackPixelsAndWidthToTurningPower(xOffset, width));
                 telemetry.addData("yMotorPower", coneStackWidthMotorPower(width));
+                telemetry.addData("width",width);
```
Comment (malavika1) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java:275]: https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace
```suggestion
                telemetry.addData("width", width);
```

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java [hunk 6]
```diff
@@ -248,20 +261,22 @@ public void centerConeStack(RobotCameraPipeline pipeline) {
...
             // drive forward while centering on the cone stack if contour exists
             if (width == 0) {
                 break;
             } else {
-                driveWithIMU(coneStackPixelsMotorPower(xOffset), coneStackWidthMotorPower(width), 0.0);
-                telemetry.addData("xMotorPower", coneStackPixelsMotorPower(xOffset));
+                //driveWithIMU(coneStackPixelsAndWidthToStrafingPower(xOffset, width), coneStackWidthMotorPower(width), 1.09*coneStackPixelsAndWidthToTurningPower(xOffset, width));
+                telemetry.addData("xStrafingPower", coneStackPixelsAndWidthToStrafingPower(xOffset, width));
+                telemetry.addData("xTurningPower", coneStackPixelsAndWidthToTurningPower(xOffset, width));
                 telemetry.addData("yMotorPower", coneStackWidthMotorPower(width));
```
Comment (malavika1) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java:274]: If you're going to call these methods in your code and in your telemetry, it's better to use a variable for speed if the method is slow and to ensure that you are actually printing out the same value that you are using in your code.

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java [hunk 7]
```diff
@@ -286,12 +301,23 @@ public double coneStackWidthMotorPower(double coneStackWidth) {
...
     /**
-     * calculates the motor power for the robot drivetrain based on the pixel offset from the cone stack
+     * calculates the turning power for the robot drivetrain based on the pixel offset from the cone stack and the width of the cone stack
      * @param pixelOffset how far cone stack is horizontally from center of camera field of view in pixels
-     * @return motor power for robot drivetrain
+     * @param width how wide the cone stack is in the camera frame in pixels
+     * @return turning power for robot drivetrain
+     */
+    public double coneStackPixelsAndWidthToTurningPower(double pixelOffset, double width) {
+        return (Constants.CONE_STACK_CENTERING_PROPORTIONAL_KP * pixelOffset)/((1.0/250.0)*width);
```
Comment (malavika1) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java:310]: 1.0/250.0 appears to have come out of thin air. Please use a variable with a clear name so you understand why you did this later. Also why does this one use 1.0/250.0 and the other one uses 1.0/200.0? Shouldn't they be the same?

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java (hunk 1 above)
Comment (malavika1) [reply Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java:320]: Yes. People need to understand where these numbers are coming from and what they mean. Using a variable with a clear name is really helpful. Also why is this different from the other one? Should they both be the same?

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/testclasses/PipelineSwitchingTest.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import org.openftc.easyopencv.OpenCvCamera;
+import org.openftc.easyopencv.OpenCvCameraFactory;
+import org.openftc.easyopencv.OpenCvCameraRotation;
+
+@Autonomous(name = "PipelineSwitchingTest")
+public class PipelineSwitchingTest extends BaseAutonomous
+{
+    @Override
+    public void runOpMode() throws InterruptedException
+    {
```
Comment (malavika1) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/testclasses/PipelineSwitchingTest.java:19]: https://google.github.io/styleguide/javaguide.html#s4.1-braces
```suggestion
public class PipelineSwitchingTest extends BaseAutonomous {
    @Override
    public void runOpMode() throws InterruptedException {
```

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/testclasses/PipelineSwitchingTest.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        // creates pipelines
+        GrabberCameraPipeline grabberCameraPipeline = new GrabberCameraPipeline();
+        RobotCameraPipeline robotCameraPipeline = new RobotCameraPipeline();
+        grabberCameraPipeline.setRanges(Constants.LOWER_YELLOW, Constants.UPPER_YELLOW);
+        robotCameraPipeline.setRanges(Constants.LOWER_BLUE, Constants.UPPER_BLUE);
+        // starts streaming camera
+        startCameraWithPipeline(grabberCameraPipeline, robotCamera, Constants.CAMERA_X, Constants.CAMERA_Y);
+        waitForStart();
+        robotCamera.closeCameraDevice();
+        while (opModeIsActive()){
```
Comment (malavika1) [inline Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/testclasses/PipelineSwitchingTest.java:33]: https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace
```suggestion
        while (opModeIsActive()) {
```

**Review by malavika1 — COMMENTED** [review]
Please update your pull request name to be a short summary of what the code changes do

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java (hunk 7 above)
Comment (SJJCoding) [reply Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java]: Resolved, added constants for both

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java (hunk 6 above)
Comment (SJJCoding) [reply Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java]: Resolved, made them use the same variable

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java (hunk 4 above)
Comment (SJJCoding) [reply Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java]: Resolved, uncommented code

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java (hunk 3 above)
Comment (SJJCoding) [reply Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java]: Resolved, added back the old constant

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java (hunk 2 above)
Comment (SJJCoding) [reply Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java]: Resolved, removed comment

File: Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java (hunk 1 above)
Comment (SJJCoding) [reply Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java]: resolved, added a constant

**Pull request closed without merging**